### DISCUSSION TITLE
Improve Event Syscall API

### DISCUSSION THREAD

**Discussion by Stebalien** [filecoin-project/FIPs#775] [Enhancements - Technical]
Currently, events are emitted as CBOR objects. Unfortunately, this makes it hard to accurately charge gas for parsing. We _currently_ handle this by tuning our gas model for EVM events but that's obviously not going to cut it.

**Proposal:**

Emit three buffers:

- `entries`: An array of `[(flags: u64, codec: u64, key_len: u32, value_len: u32), ...]`  (packed). I.e., 24 bytes per event.
- `keys`: A buffer of all keys, concatenated.
- `values`: A buffer of all values, concatenated.

The FVM will:

1. Compute the number of entries as `entries.len() / 24` (or return an error if not a multiple of 24).
2. Charge gas based on the number of entries, the length of `keys`, and the length of `values`.
3. Reconstruct the events from these parameters.

Importantly, this means the FVM can accurately charge gas before parsing anything.

**Implementation:**

This has already been implemented in https://github.com/filecoin-project/ref-fvm/pull/1807/files and merged into a "next" branch in the ref-fvm (it's consensus breaking so we don't want it in master).

**Comment by anorth** [comment]
Sounds fine to me.
